cs4660/graph/graph.py

```diff
@@ -0,0 +1,293 @@
+
+def construct_graph_from_file(graph, file_path):
+    """
+    TODO: read content from file_path, then add nodes and edges to graph object
+
+    note that grpah object will be either of AdjacencyList, AdjacencyMatrix or ObjectOriented
+
+    In example, you will need to do something similar to following:
+
+    1. add number of nodes to graph first (first line)
+    2. for each following line (from second line to last line), add them as edge to graph
+    3. return the graph
+    """
+
+    f = open(file_path)
+    data = f.read()
+    data = data.split('\n')
+   # print (data)
+    for i in range(len(data)):
+        if (i == 0):
+          #  if(isinstance(graph, AdjacencyMatrix)):
+           #     graph.initialize(int(data[i]))
+            continue
+        row = data[i].split(':')
+        graph.add_node(Node(int(row[0])))
+        graph.add_node(Node(int(row[1])))
+        a = Node(int(row[0]))
+        b = Node(int(row[1]))
+        x = Edge(a,b,int(row[2]))
+        graph.add_edge(x)
+
+    return graph
+
+class Node(object):
+    """Node represents basic unit of graph"""
+    def __init__(self, data):
+        self.data = data
+
+    def __str__(self):
+
+        return 'Node({})'.format(self.data)
+    def __repr__(self):
+
+        return 'Node({})'.format(self.data)
+
+    def __eq__(self, other_node):
+
+        return self.data == other_node.data
+
+    def __ne__(self, other):
+
+        return not self.__eq__(other)
+
+    def __hash__(self):
+
+        return hash(self.data)
+
+class Edge(object):
+    """Edge represents basic unit of graph connecting between two edges"""
+    def __init__(self, from_node, to_node, weight):
+        self.from_node = from_node
+        self.to_node = to_node
+        self.weight = weight
+    def __str__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.from_node, self.to_node, self.weight)
+    def __repr__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.from_node, self.to_node, self.weight)
+
+    def __eq__(self, other_node):
+        return self.from_node == other_node.from_node and self.to_node == other_node.to_node and self.weight == other_node.weight
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.from_node, self.to_node, self.weight))
+
+
+class AdjacencyList(object):
+    """
+    AdjacencyList is one of the graph representation which uses adjacency list to
+    store nodes and edges
+    """
+    def __init__(self):
+        self.adjacency_list = {}
+
+    def adjacent(self, node_1, node_2):
+        if (node_2 in self.adjacency_list[node_1]):
+            return True
+        else:
+            return False
+
+    def neighbors(self, node):
+       return list(self.adjacency_list[node])
+
+    def add_node(self, node):
+        if (node in self.adjacency_list):
+            return False
+        else:
+            self.adjacency_list[node] = {}
```

Comment:
This isn't really adjacency list, you have a data structure of a dictionary with a dictionary again. That is --

```
{
  Node(0): {
    Node(1): 1
  }
}
```

In other word, from this structure, your runtime analysis will also be wrong based on your assumption of the value being a list.